src/typescript/statefull.tsx

```diff
@@ -0,0 +1,276 @@
+        }}
+      >
+        +
+      </button>
+    </div>
+  );
+}
+
+
+
+
+function Gallery() {
+  const [images, setImages] = useState([
+    "https://i.ytimg.com/vi/MCn9lL94sxQ/maxresdefault.jpg",
+  ]);
+  const [numImages, setNumImages] = useState(images.length);
+  const num = useRef(0); // just playing with useRef, nothing important...
+  useEffect(() => {
+    function foo(word:string) {
+      console.log(word);
+    }
+    foo("hi ");
+    return () => {
+      foo("bye ");
+    };
+  }, [numImages]);
+
+  const handleClick = () => {
+    let src :string = prompt("new img") || ""; 
+    const imagesNew: string[] = [...images, src];
+    setImages(imagesNew);
+    setNumImages((prevNumImages) => prevNumImages + 1);
+  };
+  return (
+    <div>
+      <h2
+        onClick={() => {
+          num.current = num.current + 1;
+        
+        }}
+      >
+        Tamir
+      </h2>
+      <h1
+        onClick={() => {
+          setNumImages((prev) => prev + 1);
+        }}
+      >
+        {numImages}
+      </h1>
+      <SimpleGallery images={images}/>
+      <button onClick={handleClick}>Add </button>
+    </div>
+  );
+}
+
+interface IState { 
+    data :string[][],
+
+}
+class Canvasv2 extends Component<{},IState> {
+  cycleColors = ["red", "yellow", "blue", "orange"];
+  cycleData = [
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+  ];
+  state = {
+    data: [
+      ["red", "white", "gray", "orange"],
+      ["green", "yellow", "gray", "purple"],
+      ["red", "white", "yellow", "green"],
+      ["red", "green", "gray", "purple"],
+    ],
+  };
+  render() {
+    return (
+      <div className="simple-canvas ">
+        {this.state.data.map((row, rowIndex) => (
+          <div className="row" key={rowIndex}>
+            {row.map((color, colIndex) => (
+              <div
+                key={colIndex}
+                className="cell"
+                style={{ backgroundColor: color }}
+                onClick={() => {
+                  const newData = JSON.parse(JSON.stringify(this.state.data)); // deepcopy
+                  newData[rowIndex][colIndex] = this.cycleColors[
+                    this.cycleData[rowIndex][colIndex]
+                  ]; // update color
+                  this.cycleData[rowIndex][colIndex] =
+                    (this.cycleData[rowIndex][colIndex] + 1) %
+                    this.cycleColors.length;
+                  this.setState({ data: newData });
+                }}
+              ></div>
+            ))}
+          </div>
+        ))}
+      </div>
+    );
+  }
+}
+
+
+interface TodoProps  {
+  };
+interface ListItem{
+    title:string,
+    done:boolean,
+}
+interface TodoState  {
+    todos:ListItem[] ,
+    addInput:string,
+    searchInput:string,
+    filterValue:FilterEnum
+    
+  };
+  enum FilterEnum  { // gave valies to each member for comparing them in the future
+    all="all",
+    completed="completed",
+    incompleted = "incompleted",
+   
```

Comment:
I believe the more widely used convention is to write enum with ALL_CAPS, meaning:
```suggestion
    ALL = "ALL",
    COMPLETED = "COMPLETED",
    INCOMPLETED = "INCOMPLETED",
```

Reply:
In the left side I agree, but in the right side I want it to be according to the user select options, to make the code cleaner